hack/generator/pkg/conversions/readable_conversion_endpoint_set.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package conversions
+
+import (
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+// ReadableConversionEndpointSet is a set of uniquely named readable conversion endpoints
+type ReadableConversionEndpointSet map[string]*ReadableConversionEndpoint
+
+// NewReadableConversionEndpointSet returns a new set of readable conversion endpoints
+func NewReadableConversionEndpointSet() ReadableConversionEndpointSet {
+	return make(ReadableConversionEndpointSet)
+}
+
+// CreatePropertyEndpoints will create readable conversion endpoints for any properties found on the passed instance
+// type. Existing endpoints won't be overwritten. Returns the count of new endpoints created
+func (set ReadableConversionEndpointSet) CreatePropertyEndpoints(
+	instance astmodel.Type,
+	knownLocals *astmodel.KnownLocalsSet) int {
+	// Add an endpoint for each property we can read
+	return set.addForEachProperty(instance, func(prop *astmodel.PropertyDefinition) *ReadableConversionEndpoint {
+		return NewReadableConversionEndpointReadingProperty(prop.PropertyName(), prop.PropertyType(), knownLocals)
+	})
+}
+
+// CreateValueFunctionEndpoints will create additional readable conversion endpoints for any compatible functions found
+// on the passed instance type that don't collide with the names of existing endpoints. Returns the count of new
+// endpoints created
+func (set ReadableConversionEndpointSet) CreateValueFunctionEndpoints(
+	instance astmodel.Type,
+	knownLocals *astmodel.KnownLocalsSet) int {
+	// Add more endpoints for any value functions we can read
+	return set.addForEachValueFunction(instance, func(fn astmodel.ValueFunction) *ReadableConversionEndpoint {
+		return NewReadableConversionEndpointReadingValueFunction(fn.Name(), fn.ReturnType(), knownLocals)
+	})
+}
+
+// CreateBagItemEndpoints will create additional property bag item endpoints for any property on the passed instance
+// type that doesn't already have one. Returns the count of new endpoints created.
+func (set ReadableConversionEndpointSet) CreateBagItemEndpoints(
+	instance astmodel.Type,
+	knownLocals *astmodel.KnownLocalsSet) int {
+	// Add a property bag item endpoint for each property we don't already support
+	return set.addForEachProperty(instance, func(prop *astmodel.PropertyDefinition) *ReadableConversionEndpoint {
+		name := string(prop.PropertyName())
+		return NewReadableConversionEndpointReadingPropertyBagMember(name, prop.PropertyType(), knownLocals)
+	})
```

Comment:
I got very confused reading this and the definition of `addForEachProperty`, since the latter explicitly skips property bags - the thing I was missing was that this gets called with the *other* type after `CreatePropertyEndpoints` is called with this type, and relies on `addForEachProperty` skipping things that are already present.

Reply:
Added more comments explaining why we do this.